File: src/main/java/io/vproxy/vfx/entity/input/Key.java
package io.vproxy.vfx.entity.input;

import javafx.scene.input.MouseButton;
import vjson.JSON;
import vjson.JSONObject;
import vjson.deserializer.rule.BoolRule;
import vjson.deserializer.rule.ObjectRule;
import vjson.deserializer.rule.Rule;
import vjson.deserializer.rule.StringRule;
import vjson.util.ObjectBuilder;

import java.util.*;

public class Key implements JSONObject {
    public final MouseButton button;
    public final MouseWheelScroll scroll;
    public final KeyCode key;
    public final boolean isLeftKey;
    public final String raw;

    public static final Rule<Key> rule = ObjectRule.builder(KeyBuilder::new, KeyBuilder::build, builder -> builder
        .put("button", (o, it) -> o.button = MouseButton.valueOf(it), StringRule.get())
        .put("scroll", (o, it) -> o.scroll = it, MouseWheelScroll.rule)
        .put("key", (o, it) -> o.key = KeyCode.valueOf(it), StringRule.get())
        .put("isLeftKey", (o, it) -> o.isLeftKey = it, BoolRule.get())
        .put("raw", (o, it) -> o.raw = it, StringRule.get())
    );

    private static class KeyBuilder {
        MouseButton button;
        MouseWheelScroll scroll;
        KeyCode key;
        boolean isLeftKey;
        String raw;

        Key build() {
            if (button != null)
                return new Key(button);
            if (scroll != null)
                return new Key(scroll);
            if (key != null)
                return new Key(key, isLeftKey);
            return new Key(raw);
        }
    }

    public Key(String raw) {
        this.raw = raw;
        this.button = formatButton(raw);
        this.scroll = null;
        this.key = formatKey(raw);
        this.isLeftKey = checkIsLeftKey(raw);
    }

    public Key(MouseButton button) {
        this.button = button;
        this.scroll = null;
        this.key = null;
        this.isLeftKey = false;
        this.raw = toString(button, null, null, false);
    }

    public Key(MouseWheelScroll scroll) {
        this.button = null;
        this.scroll = scroll;
        this.key = null;
        this.isLeftKey = false;
        this.raw = toString(null, scroll, null, false);
    }

    public Key(KeyCode key) {
        this.button = null;
        this.scroll = null;
        this.key = key;
        this.isLeftKey = false;
        this.raw = toString(null, null, key, false);
    }

    public Key(KeyCode key, boolean isLeftKey) {
        this.button = null;
        this.scroll = null;
        this.key = key;
        this.isLeftKey = isLeftKey;
        this.raw = toString(null, null, key, isLeftKey);
    }

    public boolean isValid() {
        return toString(button, scroll, key, isLeftKey) != null;
    }

    private static MouseButton formatButton(String raw) {
        return stringToMouseButtonMap.get(raw.toLowerCase());
    }

    private static KeyCode formatKey(String raw) {
        return stringToKeyCodeMap.get(raw.toLowerCase());
    }

    private static boolean checkIsLeftKey(String raw) {
        raw = raw.toLowerCase();
        return hasLeftRightStrings.contains(raw) && raw.startsWith("left");
    }

    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
    private static final Map<String, MouseButton> stringToMouseButtonMap;
    private static final Map<MouseButton, String> mouseButtonToStringMap = new HashMap<>() {{
        put(MouseButton.PRIMARY, "LeftMouseButton");
        put(MouseButton.SECONDARY, "RightMouseButton");
        put(MouseButton.MIDDLE, "MiddleMouseButton");
    }};
    private static final Map<KeyCode, String> keyCodeToStringMap = new HashMap<>() {{
        for (var c : KeyCode.values()) {
            put(c, c.ueText);
        }
    }};
    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
    private static final Map<String, KeyCode> stringToKeyCodeMap;
    private static final Set<KeyCode> requireLeftRightKeys = new HashSet<>() {{
        add(KeyCode.CONTROL);
        add(KeyCode.SHIFT);
        add(KeyCode.ALT);
    }};
    private static final Set<String> hasLeftRightStrings = new HashSet<>() {{
        add("LeftControl".toLowerCase());
        add("RightControl".toLowerCase());
        add("LeftAlt".toLowerCase());
        add("RightAlt".toLowerCase());
        add("LeftShift".toLowerCase());
        add("RightShift".toLowerCase());
    }};

    static {
        stringToMouseButtonMap = new HashMap<>() {{
            for (var entry : mouseButtonToStringMap.entrySet()) {
                put(entry.getValue().toLowerCase(), entry.getKey());
            }
        }};
        stringToKeyCodeMap = new HashMap<>() {{
            for (var entry : keyCodeToStringMap.entrySet()) {
                var k = entry.getKey();
                var s = entry.getValue().toLowerCase();
                if (requireLeftRightKeys.contains(k)) {
                    put("left" + s, k);
                    put("right" + s, k);
                } else {
                    put(s, k);
                }
            }
        }};
    }

    private String toString(MouseButton button, MouseWheelScroll scroll, KeyCode key, boolean isLeftKey) {
        String ret = null;
        if (button != null) {
            ret = mouseButtonToStringMap.get(button);
        } else if (scroll != null) {
            ret = scroll.toString();
        } else if (key != null) {
            ret = keyCodeToStringMap.get(key);
            if (ret != null) {
                if (requireLeftRightKeys.contains(key)) {
                    ret = (isLeftKey ? "Left" : "Right") + ret;
                }
            }
        }
        return ret;
    }

    @Override
    public String toString() {
        String ret = toString(button, scroll, key, isLeftKey);
        if (ret == null) {
            ret = raw;
        }
        if (ret == null) {
            ret = "Unknown";
        }
        return ret;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Key key1 = (Key) o;

        if (isLeftKey != key1.isLeftKey) return false;
        if (button != key1.button) return false;
        if (key != key1.key) return false;
        return Objects.equals(raw, key1.raw);
    }

    @Override
    public int hashCode() {
        int result = button != null ? button.hashCode() : 0;
        result = 31 * result + (key != null ? key.hashCode() : 0);
        result = 31 * result + (isLeftKey ? 1 : 0);
        result = 31 * result + (raw != null ? raw.hashCode() : 0);
        return result;
    }

    @Override
    public JSON.Object toJson() {
        var ob = new ObjectBuilder();
        if (button != null) {
            ob.put("button", button.name());
        }
        if (scroll != null) {
            ob.putInst("scroll", scroll.toJson());
        }
        if (key != null) {
            ob.put("key", key.name());
            ob.put("isLeftKey", isLeftKey);
        }
        if (button == null && key == null && raw != null) {
            ob.put("raw", raw);
        }
        return ob.build();
    }
}

File: src-ts/src/entity/input/Key.ts
import { MouseWheelScroll, MouseWheelDirection } from './MouseWheelScroll.js';
import { KeyCode, keyCodeByUeText } from './KeyCode.js';

export type MouseButton = 'PRIMARY' | 'SECONDARY' | 'MIDDLE' | 'NONE';

export const MouseButton = {
  PRIMARY: 'PRIMARY' as MouseButton,
  SECONDARY: 'SECONDARY' as MouseButton,
  MIDDLE: 'MIDDLE' as MouseButton,
  NONE: 'NONE' as MouseButton,
};

const mouseButtonToString = new Map<MouseButton, string>([
  ['PRIMARY', 'LeftMouseButton'],
  ['SECONDARY', 'RightMouseButton'],
  ['MIDDLE', 'MiddleMouseButton'],
]);
const stringToMouseButton = new Map<string, MouseButton>();
for (const [k, v] of mouseButtonToString) stringToMouseButton.set(v.toLowerCase(), k);

const requireLeftRightUeTexts = new Set<string>(['shift', 'control', 'alt']);

export class Key {
  readonly button: MouseButton | null;
  readonly scroll: MouseWheelScroll | null;
  readonly key: KeyCode | null;
  readonly isLeftKey: boolean;
  readonly raw: string;

  constructor(raw: string);
  constructor(button: MouseButton, _marker?: undefined);
  constructor(scroll: MouseWheelScroll);
  constructor(key: KeyCode, isLeftKey?: boolean);
  constructor(arg0: string | MouseButton | MouseWheelScroll | KeyCode, isLeftKey?: boolean) {
    if (typeof arg0 === 'string') {
      this.raw = arg0;
      this.button = formatButton(arg0);
      this.scroll = formatScroll(arg0);
      this.key = formatKey(arg0);
      this.isLeftKey = checkIsLeftKey(arg0);
      return;
    }
    if (typeof arg0 === 'string' && (arg0 === 'PRIMARY' || arg0 === 'SECONDARY' || arg0 === 'MIDDLE' || arg0 === 'NONE')) {
      // unreachable due to TypeScript overload, but kept for runtime safety
      this.button = arg0 as MouseButton;
      this.scroll = null;
      this.key = null;
      this.isLeftKey = false;
      this.raw = render(null, null, null, false) ?? arg0;
      return;
    }
    if (arg0 instanceof MouseWheelScroll) {
      this.button = null;
      this.scroll = arg0;
      this.key = null;
      this.isLeftKey = false;
      this.raw = render(null, arg0, null, false) ?? 'Unknown';
      return;
    }
    const kc = arg0 as KeyCode;
    const left = isLeftKey ?? false;
    this.button = null;
    this.scroll = null;
    this.key = kc;
    this.isLeftKey = left;
    this.raw = render(null, null, kc, left) ?? 'Unknown';
  }

  isValid(): boolean {
    return render(this.button, this.scroll, this.key, this.isLeftKey) !== null;
  }

  toString(): string {
    const r = render(this.button, this.scroll, this.key, this.isLeftKey);
    return r ?? this.raw ?? 'Unknown';
  }

  equals(o: unknown): boolean {
    if (!(o instanceof Key)) return false;
    return (
      o.isLeftKey === this.isLeftKey &&
      o.button === this.button &&
      o.key === this.key &&
      o.raw === this.raw
    );
  }

  hashCode(): number {
    let h = 0;
    if (this.button) h = mix(h, hashStr(this.button));
    if (this.key) h = mix(h, hashStr(this.key.ueText));
    h = mix(h, this.isLeftKey ? 1 : 0);
    if (this.raw) h = mix(h, hashStr(this.raw));
    return h;
  }

  toJson(): Record<string, unknown> {
    const obj: Record<string, unknown> = {};
    if (this.button) obj.button = this.button;
    if (this.scroll) obj.scroll = this.scroll.toJson();
    if (this.key) {
      obj.key = this.key.name;
      obj.isLeftKey = this.isLeftKey;
    }
    if (!this.button && !this.key && this.raw) obj.raw = this.raw;
    return obj;
  }
}

function formatButton(raw: string): MouseButton | null {
  return stringToMouseButton.get(raw.toLowerCase()) ?? null;
}

function formatScroll(raw: string): MouseWheelScroll | null {
  if (!raw.startsWith('scroll-')) return null;
  let r = raw.substring('scroll-'.length);
  let dir: string;
  let val: string;
  if (r.includes(':')) {
    const split = r.split(':');
    if (split.length !== 2) return null;
    dir = split[0]!;
    val = split[1]!;
  } else {
    dir = r;
    val = '0';
  }
  const d = dir.toUpperCase() as MouseWheelDirection;
  if (d !== MouseWheelDirection.UP && d !== MouseWheelDirection.DOWN) return null;
  const v = parseInt(val, 10);
  if (Number.isNaN(v)) return null;
  return new MouseWheelScroll(d, v);
}

function formatKey(raw: string): KeyCode | null {
  const direct = keyCodeByUeText(raw);
  if (direct) return direct;
  // Java's stringToKeyCodeMap maps both "leftshift" and "rightshift" to SHIFT
  // (and similarly for control/alt). Strip the Left/Right prefix for modifier keys.
  const lower = raw.toLowerCase();
  for (const prefix of ['left', 'right']) {
    if (lower.startsWith(prefix)) {
      const stripped = lower.substring(prefix.length);
      const kc = keyCodeByUeText(stripped);
      if (kc) return kc;
    }
  }
  return null;
}

function checkIsLeftKey(raw: string): boolean {
  const r = raw.toLowerCase();
  return (
    requireLeftRightUeTexts.has(r.replace(/^left|^right/, '')) ||
    (r.startsWith('left') && /left(shift|control|alt)/.test(r))
  ) && r.startsWith('left');
}

function render(
  button: MouseButton | null,
  scroll: MouseWheelScroll | null,
  key: KeyCode | null,
  isLeftKey: boolean,
): string | null {
  if (button !== null) return mouseButtonToString.get(button) ?? null;
  if (scroll !== null) return scroll.toString();
  if (key !== null) {
    let ret = key.ueText;
    if (requireLeftRightUeTexts.has(key.ueText.toLowerCase())) {
      ret = (isLeftKey ? 'Left' : 'Right') + ret;
    }
    return ret;
  }
  return null;
}

function mix(h: number, v: number): number {
  return ((31 * h) + v) | 0;
}
function hashStr(s: string): number {
  let h = 0;
  for (let i = 0; i < s.length; i++) h = (31 * h + s.charCodeAt(i)) | 0;
  return h;
}
